File: 239-sliding-window-maximum/239-sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        
        deque<int> d;
        
        int i=0, j=0;
        
        vector<int> res;
        
        while(j< nums.size()){
            
            while(d.size() > 0 && d.back() < nums[j]){
                d.pop_back();
            }
            
            d.push_back(nums[j]);
            
            if(j-i+1<k) j++;
            
            else if(j-i+1 == k){
                
                res.push_back(d.front());
                
                if(nums[i] == d.front()){
                    d.pop_front();
                }
                
                i++;
                j++;
            }
        }
        
        return res;
    }
};

File: 239-sliding-window-maximum/239-sliding-window-maximum.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        
        output = []
        q = collections.deque()
        
        l=r=0
        
        while r < len(nums):
            
            while q and q[-1] < nums[r]:
                q.pop()
                
            q.append(nums[r])
            
            if r-l+1 < k:
                r += 1
            
            elif r-l+1 == k:
                output.append(q[0])
                if nums[l] == q[0]:
                    q.popleft()
                    
                l += 1
                r += 1
        return output